pytorch_lightning/trainer/training_loop.py

```diff
@@ -679,20 +680,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
                     # -------------------
                     # calculate loss (train step + train step end)
                     # -------------------
+
+                    # no ddp sync at the beginning of forward or backward due to parameter changes in this or last step required
+                    no_sync = self._updated_model_last_step and isinstance(self.trainer.model, torch.nn.parallel.DistributedDataParallel)
```

Comment:
Great catch there !

Would you mind making this a bit more readable ? 
Like moving no_sync outside of main loop and hiding those __enter__() and __exit__() ?
Also, you don't need `self._updated_model_last_step` variable. It can derived from `self.trainer.batch_idx and self.trainer.accumulate_grad_batches`

Reply:
@tchaton Regarding the variable you're probably right. 

Where do you want to move it and how do you want to hide those enter and exit methods?
I haven't found a better place for it, since it affects every DDP backend we have so far.